### DISCUSSION TITLE
Custom Path/URL For Pages

### DISCUSSION THREAD

**Discussion by 4lch4** [squidfunk/mkdocs-material#4693] [Q&A]
Hey folks!

I posted my question [on Twitter](https://twitter.com/4lch4/status/1597325011957997568) initially but @squidfunk recommended I post here instead so here it goes.

I've put together a sample repository that shows exactly what I'm after: https://github.com/4lch4/MkDocs-Sample

Basically, if you use the above sample, how do I make it so the `_index.md` file located in the `/docs/notes` directory is accessible at `localhost:8000/notes/index` instead of the default `localhost:8000/notes/_index`.

I'd be more than happy to provide more information if needed, just let me know ^_^

**Comment by squidfunk** [comment]
The most important thing: why would you want to do that? For what I know, no, it's not possible, but if we know what you're trying to achieve, there's probably a different way.

**Reply by 4lch4** [reply]
Mostly because it's a convention I follow where index files start with an underscore to ensure it's at the top of the list, but having it in the URL can be kind of annoying.

**Reply by squidfunk** [reply]
While there is no off-the-shelf solution I know off, you could write a [custom hook](https://www.mkdocs.org/user-guide/configuration/#hooks) (new in MkDocs 1.4) that changes the destination path and URL of the resulting page. [We're doing something similar in our blog plugin](https://github.com/squidfunk/mkdocs-material-insiders/blob/78de2dfe6050c4354154224fad920c05e25cef1b/src/plugins/blog/plugin.py#L249-L254). Maybe something like (untested):

``` py
def on_page_markdown(markdown, *, page, config, files):
    if page.file.name == "_index":
        page.file.name = "index"
        page.file.url = page.file.url.replace("_index", "index")
        page.file.dest_uri = page.file.dest_uri.replace("_index", "index")
        page.file.abs_dest_path = page.file.abs_dest_path.replace("_index", "index")
```

**Reply by 4lch4** [reply]
Ooo... I'll tinker with this today/this weekend and report back. Thanks a lot, Squidfunk!

**Reply by 4lch4** [reply]
Alrighty, I tried this out but got an error that I've pasted below.

```bash
jisodl0 at MACX-382FQ05P in ~/Development/Playground/MkDocs-Sample (main●●) 
$ mkdocs serve
zsh: /Users/jisodl0/Library/Python/3.8/bin/mkdocs: bad interpreter: /Library/Developer/CommandLineTools/usr/bin/python3: no such file or directory
INFO     -  Building documentation...
INFO     -  Cleaning site directory
ERROR    -  Error reading page 'index.md': on_page_markdown() missing 1 required positional argument: 'markdown'
Traceback (most recent call last):
  File "/opt/homebrew/bin/mkdocs", line 8, in <module>
    sys.exit(cli())
  File "/opt/homebrew/lib/python3.9/site-packages/click-8.0.3-py3.9.egg/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/opt/homebrew/lib/python3.9/site-packages/click-8.0.3-py3.9.egg/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/opt/homebrew/lib/python3.9/site-packages/click-8.0.3-py3.9.egg/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/homebrew/lib/python3.9/site-packages/click-8.0.3-py3.9.egg/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/homebrew/lib/python3.9/site-packages/click-8.0.3-py3.9.egg/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/opt/homebrew/lib/python3.9/site-packages/mkdocs/__main__.py", line 234, in serve_command
    serve.serve(dev_addr=dev_addr, livereload=livereload, watch=watch, **kwargs)
  File "/opt/homebrew/lib/python3.9/site-packages/mkdocs/commands/serve.py", line 83, in serve
    builder(config)
  File "/opt/homebrew/lib/python3.9/site-packages/mkdocs/commands/serve.py", line 76, in builder
    build(config, live_server=live_server, dirty=dirty)
  File "/opt/homebrew/lib/python3.9/site-packages/mkdocs/commands/build.py", line 308, in build
    _populate_page(file.page, config, files, dirty)
  File "/opt/homebrew/lib/python3.9/site-packages/mkdocs/commands/build.py", line 177, in _populate_page
    page.markdown = config.plugins.run_event(
  File "/opt/homebrew/lib/python3.9/site-packages/mkdocs/plugins.py", line 520, in run_event
    result = method(item, **kwargs)
TypeError: on_page_markdown() missing 1 required positional argument: 'markdown'
```

I pushed the changes I made to the sample repo above in case it helps.

Also, am I putting the hooks property in the right location? I added it as a top-level property but the schema is saying this is an invalid use and the hooks property doesn't belong there but I can't find any other location it should go.

_EDIT: Here's a screenshot of the error regarding the `hooks` property:_

![image](https://user-images.githubusercontent.com/6931731/205326022-beb33415-32fb-4920-b5d3-0034a1de745d.png)

**Reply by squidfunk** [reply]
Two things: I had the time to actually run the example and corrected it, it should now work as expected. The other thing is that `hooks` was not defined in our schema. This is fixed in 2d4fe293f.
